src/redisearch_rs/c_entrypoint/inverted_index_ffi/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+        }
+        (FREQS_FIELDS_MASK, _, _) => {
+            InvertedIndex::FreqsFields(FieldMaskTrackingIndex::new(flags, FreqsFields))
+        }
+        (FREQS_FIELDS_WIDE_MASK, _, _) => {
+            InvertedIndex::FreqsFieldsWide(FieldMaskTrackingIndex::new(flags, FreqsFieldsWide))
+        }
+        (FREQS_ONLY_MASK, _, _) => InvertedIndex::FreqsOnly(II::new(flags, FreqsOnly)),
+        (FIELDS_ONLY_MASK, _, _) => {
+            InvertedIndex::FieldsOnly(FieldMaskTrackingIndex::new(flags, FieldsOnly))
+        }
+        (FIELDS_ONLY_WIDE_MASK, _, _) => {
+            InvertedIndex::FieldsOnlyWide(FieldMaskTrackingIndex::new(flags, FieldsOnlyWide))
+        }
+        (FIELDS_OFFSETS_MASK, _, _) => {
+            InvertedIndex::FieldsOffsets(FieldMaskTrackingIndex::new(flags, FieldsOffsets))
+        }
+        (FIELDS_OFFSETS_WIDE_MASK, _, _) => {
+            InvertedIndex::FieldsOffsetsWide(FieldMaskTrackingIndex::new(flags, FieldsOffsetsWide))
+        }
+        (OFFSETS_ONLY_MASK, _, _) => InvertedIndex::OffsetsOnly(II::new(flags, OffsetsOnly)),
+        (FREQS_OFFSETS_MASK, _, _) => InvertedIndex::FreqsOffsets(II::new(flags, FreqsOffsets)),
+        (DOC_IDS_ONLY_MASK, false, _) => InvertedIndex::DocumentIdOnly(II::new(flags, DocIdsOnly)),
+        (DOC_IDS_ONLY_MASK, true, _) => {
+            InvertedIndex::RawDocumentIdOnly(II::new(flags, RawDocIdsOnly))
+        }
+        (NUMERIC_MASK, _, false) => {
+            InvertedIndex::Numeric(EntriesTrackingIndex::new(flags, Numeric::new()))
+        }
+        (NUMERIC_MASK, _, true) => InvertedIndex::Numeric(EntriesTrackingIndex::new(
+            flags,
+            Numeric::new().with_float_compression(),
+        )),
+        _ => panic!("Unsupported index flags: {flags:?}"),
+    };
+
+    *mem_size = ii_dispatch!(&ii, memory_usage);
+
+    let ii_boxed = Box::new(ii);
+    Box::into_raw(ii_boxed)
+}
+
+/// Free the memory associated with the inverted index instance created using [`NewInvertedIndex_Ex`].
+///
+/// # Safety
+/// The following invariant must be upheld when calling this function:
+/// - `ii` must be a valid pointer to an `InvertedIndex` instance created using
+///   [`NewInvertedIndex_Ex`] or `NewInvertedIndex`.
+/// - `ii` must not be NULL.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn InvertedIndex_Free(ii: *mut InvertedIndex) {
+    // SAFETY: The caller must ensure that `ii` is a valid pointer to an `InvertedIndex`
+    let _ = unsafe { Box::from_raw(ii) };
+}
+
+/// Get the memory usage of the inverted index instance in bytes.
+///
+/// # Safety
+/// The following invariant must be upheld when calling this function:
+/// - `ii` must be a valid pointer to an `InvertedIndex` instance and must not be NULL.
+#[unsafe(no_mangle)]
+pub unsafe extern "C" fn InvertedIndex_MemUsage(ii: *const InvertedIndex) -> usize {
+    // SAFETY: The caller must ensure that `ii` is a valid pointer to an `InvertedIndex`
```

Comment:
Every time we assume the pointer must not be NULL, we should have a debug assertions that checks it.